third_party/amd/lib/TritonAMDGPUTransforms/HoistLayoutConversions.cpp

```diff
@@ -0,0 +1,63 @@
+#include "TritonAMDGPUTransforms/Passes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+
+using namespace mlir;
+namespace ttg = mlir::triton::gpu;
+
+// Hoist convert_layout out of the loop if the src is defined out of the loop.
+// This is a heuristic driven by optimizing fused attention kernels, in which
+// we want to load Q tensor and keep it in register, instead of loading it
+// (neither from global or shared memory) at every iteration of the loop.
+static void hoistCvtDotOpOutOfLoop(ttg::ConvertLayoutOp cvtOp) {
+  // Check the dst of cvt has dotOperand layout
+  RankedTensorType rtType = dyn_cast<RankedTensorType>(cvtOp.getType());
+  if (!rtType)
+    return;
+  Attribute encoding = rtType.getEncoding();
+  if (!encoding)
+    return;
+  if (!isa<ttg::DotOperandEncodingAttr>(encoding))
+    return;
+  // Check the src of cvt is defined out of the loop
+  auto srcDefOp = cvtOp.getSrc().getDefiningOp();
+  if (srcDefOp) {
+    scf::ForOp parentForOp = cvtOp->getParentOfType<scf::ForOp>();
+    if (parentForOp && !parentForOp->isAncestor(srcDefOp)) {
+      cvtOp->moveAfter(srcDefOp);
+    }
+  }
+}
+
+#define GEN_PASS_CLASSES
+#include "TritonAMDGPUTransforms/Passes.h.inc"
+
+namespace {
+struct TritonAMDGPUHoistLayoutConversionsPass
+    : public TritonAMDGPUHoistLayoutConversionsBase<
+          TritonAMDGPUHoistLayoutConversionsPass> {
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    ModuleOp m = getOperation();
+
+    for (auto funcOp : m.getOps<triton::FuncOp>()) {
+      funcOp.walk([&](ttg::ConvertLayoutOp cvtOp) -> void {
```

Comment:
You may want to first collect all convert ops into a `SmallVector` and then process them to be deterministic. Moving convert ops while walking can cause subtle issues.